src/ACL.cpp

```diff
@@ -141,36 +141,69 @@ QFlags< ChanACL::Perm > ChanACL::effectivePermissions(ServerUser *p, Channel *ch
 	bool write    = false;
 	ChanACL *acl;
 
+	// Iterate over all parent channels from root to the channel the user is in (inclusive)
 	while (!chanstack.isEmpty()) {
 		ch = chanstack.pop();
-		if (!ch->bInheritACL)
+		if (!ch->bInheritACL) {
 			granted = def;
+		}
 
 		foreach (acl, ch->qlACL) {
 			bool matchUser  = (acl->iUserId != -1) && (acl->iUserId == p->iId);
 			bool matchGroup = Group::appliesToUser(*chan, *ch, acl->qsGroup, *p);
+
+			bool applyFromSelf   = (ch == chan && acl->bApplyHere);
+			bool applyFromParent = (ch != chan && acl->bApplySubs);
+
+			// Not using applyFromAny will ignore the context flags for the current ACL
+			bool applyFromAny = applyFromSelf || applyFromParent;
```

Comment:
I suppose we can short-circuit in case `!applyFromAny` and just jump to the next loop iteration?

That should make explicit checking of this boolean further down redundant :thinking: